File: 0437.PathSumIII.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func pathSum(root *TreeNode, sum int) int {
    // O(n) + O(n)
    Map := make(map[int]int)
    Map[0] = 1
    return dfs(root, sum, 0, Map)
    /* O(n^2) ~ O(nlgn) + O(n) ~ O(lgn)
    if root == nil {
        return 0
    }
    return pathSumFrom(root, sum) + pathSum(root.Left, sum) + pathSum(root.Right, sum)
    */
}

func dfs(node *TreeNode, sum, curSum int, Map map[int]int) int {
    if node == nil {
        return 0
    }
    curSum += node.Val
    cnt := Map[curSum-sum]
    Map[curSum]++
    cnt += dfs(node.Left, sum, curSum, Map) + dfs(node.Right, sum, curSum, Map)
    Map[curSum]--
    return cnt
}

// O(n) ~ O(lgn)
func pathSumFrom(node *TreeNode, sum int) int {
    if node == nil {
        return 0
    }
    cnt := 0
    if node.Val == sum {
        cnt++
    }
    return cnt + pathSumFrom(node.Left, sum-node.Val) + pathSumFrom(node.Right, sum-node.Val)
}

File: 0000-0999/0437.PathSumIII.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: int
        """
        #fast: O(n)
        return self.dfs2(root, sum, [sum])
        
        #Slow: O(n*logn) - O(n*n)
        if not root:
            return 0
        return self.dfs(root, sum) + self.pathSum(root.left, sum) + self.pathSum(root.right, sum)
    
    def dfs2(self, root, sum, targets):
        if not root:
            return 0
        count = 0
        for t in targets:
            if root.val == t:
                count += 1
        targets = [t-root.val for t in targets] + [sum]
        count += self.dfs2(root.left, sum, targets) + self.dfs2(root.right, sum, targets)
        return count
        
    def dfs(self, root, sum):
        if not root:
            return 0
        count = 0
        if root.val == sum:
            count += 1
        count += self.dfs(root.left, sum-root.val) + self.dfs(root.right, sum-root.val)
        return count
        